lib/improver/nbhood/circular_kernel.py

```diff
@@ -172,3 +175,231 @@ def run(self, cube, radius):
             cube, radius, MAX_RADIUS_IN_GRID_CELLS)
         cube = self.apply_circular_kernel(cube, ranges)
         return cube
+
+
+class GeneratePercentilesFromACircularNeighbourhood(object):
+    """
+    Methods for use in calculating percentiles from a 2D circular
+    neighbourhood.
+    A maximum kernel radius of 500 grid cells is imposed in order to
+    avoid computational ineffiency and possible memory errors.
+    """
+    def __init__(self, percentiles=DEFAULT_PERCENTILES):
+        """
+        Initialise class.
+
+        Args:
+            percentiles : list (optional)
+                Percentile values at which to calculate; if not provided uses
+                DEFAULT_PERCENTILES.
+
+        """
+        self.percentiles = tuple(percentiles)
+
+    def __repr__(self):
+        """Represent the configured class instance as a string."""
+        result = ('<GeneratePercentilesFromACircularNeighbourhood: '
+                  'percentiles: {}>')
+        return result.format(self.percentiles)
+
+    def pad_and_unpad_cube(self, slice_2d, kernel):
+        """
+        Method to pad and unpad a two dimensional cube. The input array is
+        padded and percentiles are calculated using a neighbourhood around
+        each point. The resultingpercentile data are unpadded and put into a
+        cube.
+
+        Args:
+            slice_2d : Iris.cube.Cube
+                2d cube to be padded with a halo.
+            kernel : Numpy array
+                Kernel used to specify the neighbourhood to consider when
+                calculating the percentiles within a neighbourhood.
+
+        Examples:
```

Comment:
Great example :D now we actually have a reference for what is going on.

Reply:
The example looks a lot better if it looks like this:
```
        Examples:
            1. Take the input slice_2d cube with the data, where 1 is an
               occurrence and 0 is an non-occurrence::
                    [[1., 1., 1.,],
                     [1., 0., 1.],
                     [1., 1., 1.]]
            2. Define a kernel. This kernel is effectively placed over each
               point within the input data. Note that the input data is padded
               prior to placing the kernel over each point, so that the kernel
               does not exceed the bounds of the padded data::
                    [[ 0.,  0.,  1.,  0.,  0.],
                     [ 0.,  1.,  1.,  1.,  0.],
                     [ 1.,  1.,  1.,  1.,  1.],
                     [ 0.,  1.,  1.,  1.,  0.],
                     [ 0.,  0.,  1.,  0.,  0.]]
            3. Pad the input data. The extent of the padding is given by the
               shape of the kernel. The number of values included within the
               calculation of the mean is determined by the size of the kernel::
                    [[ 0.75,  0.75,  1.  ,  0.5 ,  1.  ,  0.75,  0.75],
                     [ 0.75,  0.75,  1.  ,  0.5 ,  1.  ,  0.75,  0.75],
                     [ 1.  ,  1.  ,  1.  ,  1.  ,  1.  ,  1.  ,  1.  ],
                     [ 0.5 ,  0.5 ,  1.  ,  0.  ,  1.  ,  0.5 ,  0.5 ],
                     [ 1.  ,  1.  ,  1.  ,  1.  ,  1.  ,  1.  ,  1.  ],
                     [ 0.75,  0.75,  1.  ,  0.5 ,  1.  ,  0.75,  0.75],
                     [ 0.75,  0.75,  1.  ,  0.5 ,  1.  ,  0.75,  0.75]]
            4. Calculate the values at the percentiles: [10].
               For the point in the upper right corner within the original
               input data e.g. ::
                    [[->1.<-, 1., 1.,],
                     [  1.,   0., 1.],
                     [  1.,   1., 1.]]
               When the kernel is placed over this point within the padded
               data, then the following points are included::
                    [[   0.75,    0.75,  ->1.<-,  0.5 ,  1.  ,  0.75,  0.75],
                     [   0.75,  ->0.75,    1.  ,  0.5<-, 1.  ,  0.75,  0.75],
                     [ ->1.  ,    1.  ,    1.  ,  1.  ,  1.<-,  1.  ,  1.  ],
                     [   0.5 ,  ->0.5 ,    1.  ,  0.<-,  1.  ,  0.5 ,  0.5 ],
                     [   1.  ,    1.  ,  ->1.<-,  1.  ,  1.  ,  1.  ,  1.  ],
                     [   0.75,    0.75,    1.  ,  0.5 ,  1.  ,  0.75,  0.75],
                     [   0.75,    0.75,    1.  ,  0.5 ,  1.  ,  0.75,  0.75]]
               This gives
               ::
                    [0, 0.5, 0.5, 0.75, 1., 1., 1., 1., 1., 1., 1., 1., 1.]
               As there are 13 points within the kernel, this gives the
               following relationship between percentiles and values.
                  ======  ==========
                  Values  Percentile
                  ======  ==========
                  0.      0
                  0.5     8.33
                  0.5     16.67
                  0.75    25.0
                  1.      33.33
                  1.      41.67
                  1.      50.0
                  1.      58.33
                  1.      66.67
                  1.      75.0
                  1.      83.33
                  1.      91.66
                  1.      100.
                  ======  ==========
               Therefore, for the 10th percentile at the value returned for
               the point in the upper right corner of the original input data
               is 0.5.
               When this process is applied to every point within the original
               input data, the result is::
                    [[[ 0.75,  0.75,  0.5 ,  0.5 ,  0.5 ,  0.75,  0.75],
                      [ 0.75,  0.55,  0.55,  0.5 ,  0.55,  0.55,  0.55],
                      [ 0.55,  0.55,  0.5 ,  0.5 ,  0.5 ,  0.5 ,  0.5 ],
                      [ 0.5 ,  0.5 ,  0.5 ,  0.5 ,  0.5 ,  0.5 ,  0.5 ],
                      [ 0.5 ,  0.5 ,  0.5 ,  0.5 ,  0.5 ,  0.55,  0.55],
                      [ 0.55,  0.55,  0.55,  0.5 ,  0.55,  0.55,  0.75],
                      [ 0.75,  0.75,  0.5 ,  0.5 ,  0.5 ,  0.75,  0.75]]],
            5. The padding is then removed to give
               ::
                   [[[ 0.5,  0.5,  0.5],
                     [ 0.5,  0.5,  0.5],
                     [ 0.5,  0.5,  0.5]]]

        """
```

Reply:
I've used your corrected docstring, which seems to display correctly.